File: 64-最小路径和.py
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
        nums = [sum(grid[m - 1][k:]) for k in range(n)]
        
        for i in range(m - 2, -1, -1):
            nums[n - 1] += grid[i][n - 1]
            for j in range(n - 2, -1, -1):
                nums[j] = grid[i][j] + min(nums[j + 1], nums[j])

        return nums[0]

File: 64-最小路径和.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        vector<int> myvector(n, grid[m - 1][n - 1]);
        for(int i = n - 2; i > -1; i--){
            myvector[i] = grid[m - 1][i] + myvector[i + 1];
        }

        for(int i = m - 2; i > -1; i--){
            myvector[n - 1] += grid[i][n - 1];
            for(int j = n - 2; j > -1; j--){
                myvector[j] = grid[i][j] + min(myvector[j + 1], myvector[j]);
            }
        }

        return myvector[0];
    }
};
